File: frontend/src/types/Gift.ts
// Import Contact interface from the existing contacts service
import { Contact } from '../services/contacts';

// Gift status and priority enums matching backend
export type GiftStatus = 'idea' | 'planned' | 'purchased' | 'wrapped' | 'given' | 'returned';
export type GiftPriority = 1 | 2 | 3 | 4; // LOW, MEDIUM, HIGH, URGENT

// Core Gift interface
export interface Gift {
  id: number;
  tenant_id: number;
  user_id: number;
  title: string;
  description?: string;
  category?: string;
  recipient_contact_id?: number;
  recipient_name?: string;
  occasion?: string;
  occasion_date?: string; // ISO string
  budget_amount?: number;
  actual_amount?: number;
  currency: string;
  status: GiftStatus;
  priority: GiftPriority;
  store_name?: string;
  purchase_url?: string;
  purchase_date?: string; // ISO string
  gift_details: Record<string, any>;
  tracking_number?: string;
  delivery_date?: string; // ISO string
  notes?: string;
  image_url?: string;
  reminder_dates: Record<string, string>; // JSON with reminder types and dates
  task_id?: number;
  transaction_id?: number;
  is_active: boolean;
  is_surprise: boolean;
  created_at: string; // ISO string
  updated_at?: string; // ISO string
}

// Gift creation interface
export interface GiftCreate {
  title: string;
  description?: string;
  category?: string;
  recipient_contact_id?: number;
  recipient_name?: string;
  occasion?: string;
  occasion_date?: string; // ISO string
  budget_amount?: number;
  currency?: string;
  status?: GiftStatus;
  priority?: GiftPriority;
  store_name?: string;
  purchase_url?: string;
  purchase_date?: string; // ISO string
  gift_details?: Record<string, any>;
  tracking_number?: string;
  delivery_date?: string; // ISO string
  notes?: string;
  image_url?: string;
  reminder_dates?: Record<string, string>;
  is_surprise?: boolean;
}

// Gift update interface
export interface GiftUpdate {
  title?: string;
  description?: string;
  category?: string;
  recipient_contact_id?: number;
  recipient_name?: string;
  occasion?: string;
  occasion_date?: string; // ISO string
  budget_amount?: number;
  actual_amount?: number;
  currency?: string;
  status?: GiftStatus;
  priority?: GiftPriority;
  store_name?: string;
  purchase_url?: string;
  purchase_date?: string; // ISO string
  gift_details?: Record<string, any>;
  tracking_number?: string;
  delivery_date?: string; // ISO string
  notes?: string;
  image_url?: string;
  reminder_dates?: Record<string, string>;
  is_surprise?: boolean;
}

// Gift Idea interface
export interface GiftIdea {
  id: number;
  tenant_id: number;
  user_id: number;
  title: string;
  description?: string;
  category?: string;
  target_contact_id?: number;
  target_demographic?: string;
  suitable_occasions: string[];
  price_range_min?: number;
  price_range_max?: number;
  currency: string;
  idea_details: Record<string, any>;
  source_url?: string;
  source_description?: string;
  image_url?: string;
  rating?: number; // 1-5 star rating
  notes?: string;
  times_gifted: number;
  last_gifted_date?: string; // ISO string
  tags: string[];
  is_active: boolean;
  created_at: string; // ISO string
  updated_at?: string; // ISO string
}

// Gift Idea creation interface
export interface GiftIdeaCreate {
  title: string;
  description?: string;
  category?: string;
  target_contact_id?: number;
  target_demographic?: string;
  suitable_occasions?: string[];
  price_range_min?: number;
  price_range_max?: number;
  currency?: string;
  idea_details?: Record<string, any>;
  source_url?: string;
  source_description?: string;
  image_url?: string;
  rating?: number;
  notes?: string;
  tags?: string[];
}

// Gift Idea update interface
export interface GiftIdeaUpdate {
  title?: string;
  description?: string;
  category?: string;
  target_contact_id?: number;
  target_demographic?: string;
  suitable_occasions?: string[];
  price_range_min?: number;
  price_range_max?: number;
  currency?: string;
  idea_details?: Record<string, any>;
  source_url?: string;
  source_description?: string;
  image_url?: string;
  rating?: number;
  notes?: string;
  tags?: string[];
}

// Filtering interfaces
export interface GiftFilters {
  status?: GiftStatus[];
  category?: string[];
  occasion?: string[];
  recipient_contact_id?: number[];
  min_budget?: number;
  max_budget?: number;
  currency?: string;
  is_surprise?: boolean;
  search?: string;
}

export interface GiftIdeaFilters {
  category?: string[];
  target_contact_id?: number[];
  min_rating?: number;
  min_price?: number;
  max_price?: number;
  currency?: string;
  is_favorite?: boolean;
  search?: string;
}

// Budget tracking interfaces
export interface GiftBudget {
  total_budget: number;
  spent_amount: number;
  remaining_budget: number;
  currency: string;
  period_start?: string;
  period_end?: string;
}

export interface BudgetInsight {
  category: string;
  budgeted: number;
  spent: number;
  remaining: number;
  percentage_used: number;
}

// Analytics interfaces
export interface GiftAnalytics {
  total_gifts: number;
  total_ideas: number;
  gifts_given: number;
  total_spent: number;
  average_gift_value: number;
  top_categories: Array<{ category: string; count: number }>;
  upcoming_occasions: Array<{ occasion: string; date: string; gifts_planned: number }>;
}

// Reference data interfaces
export interface GiftCategoryReference {
  name: string;
  description: string;
  typical_price_range: string;
  occasions: string[];
  icon?: string;
}

export interface GiftOccasionReference {
  name: string;
  type: 'annual' | 'milestone' | 'spontaneous';
  importance: 'low' | 'medium' | 'high';
  advance_notice_days: number[];
  typical_budget_range: string;
  fixed_date?: string; // MM-DD format
}

export interface GiftBudgetRangeReference {
  name: string;
  min_amount: number;
  max_amount: number;
  currency: string;
  description: string;
  suggested_categories: string[];
}

// Constants for UI
export const GIFT_STATUS_OPTIONS: { value: GiftStatus; label: string; color: string }[] = [
  { value: 'idea', label: 'Idea', color: 'gray' },
  { value: 'planned', label: 'Planned', color: 'blue' },
  { value: 'purchased', label: 'Purchased', color: 'yellow' },
  { value: 'wrapped', label: 'Wrapped', color: 'purple' },
  { value: 'given', label: 'Given', color: 'green' },
  { value: 'returned', label: 'Returned', color: 'red' },
];

export const GIFT_PRIORITY_OPTIONS: { value: GiftPriority; label: string; color: string }[] = [
  { value: 1, label: 'Low', color: 'green' },
  { value: 2, label: 'Medium', color: 'yellow' },
  { value: 3, label: 'High', color: 'orange' },
  { value: 4, label: 'Urgent', color: 'red' },
];

export const GIFT_CATEGORIES = [
  'Electronics',
  'Clothing',
  'Books',
  'Jewelry',
  'Home & Garden',
  'Sports & Outdoors',
  'Food & Beverages',
  'Health & Beauty',
  'Toys & Games',
  'Art & Crafts',
  'Music & Media',
  'Travel',
  'Experience',
  'Other'
];

export const GIFT_OCCASIONS = [
  'Birthday',
  'Christmas',
  'Anniversary',
  'Valentine\'s Day',
  'Mother\'s Day',
  'Father\'s Day',
  'Graduation',
  'Wedding',
  'Housewarming',
  'Baby Shower',
  'Retirement',
  'Promotion',
  'Just Because',
  'Thank You',
  'Apology',
  'Other'
];

// Gift form data interface
export interface GiftFormData {
  title: string;
  description: string;
  category: string;
  recipient_contact_id: number | null;
  recipient_name: string;
  occasion: string;
  occasion_date: string;
  budget_amount: string;
  currency: string;
  status: GiftStatus;
  priority: GiftPriority;
  store_name: string;
  purchase_url: string;
  gift_details: Record<string, any>;
  notes: string;
  is_surprise: boolean;
}

export interface GiftIdeaFormData {
  title: string;
  description: string;
  category: string;
  target_contact_id: number | null;
  target_demographic: string;
  suitable_occasions: string[];
  price_range_min: string;
  price_range_max: string;
  currency: string;
  source_url: string;
  rating: number;
  notes: string;
  tags: string[];
}

// Validation interfaces
export interface ValidationErrors {
  [key: string]: string;
}

// Sort interfaces
export interface GiftSortBy {
  field: 'title' | 'occasion_date' | 'budget_amount' | 'created_at' | 'status' | 'priority';
  direction: 'asc' | 'desc';
}

export interface GiftIdeaSortBy {
  field: 'title' | 'rating' | 'price_range_min' | 'created_at' | 'times_gifted';
  direction: 'asc' | 'desc';
}

// View mode for gift display
export type GiftViewMode = 'list' | 'grid' | 'timeline';

File: backend/app/models/gift.py
from sqlalchemy import Column, Integer, String, Boolean, DateTime, ForeignKey, Text, Date, JSON, Numeric
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from app.db.base_class import Base
import enum


class GiftStatus(enum.Enum):
    IDEA = "idea"
    PLANNED = "planned"
    PURCHASED = "purchased"
    WRAPPED = "wrapped"
    GIVEN = "given"
    RETURNED = "returned"


class GiftPriority(enum.IntEnum):
    LOW = 1
    MEDIUM = 2
    HIGH = 3
    URGENT = 4


class Gift(Base):
    __tablename__ = "gifts"

    id = Column(Integer, primary_key=True, index=True)
    tenant_id = Column(Integer, ForeignKey("tenants.id"), nullable=False, index=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False, index=True)
    
    # Basic Information
    title = Column(String(255), nullable=False)
    description = Column(Text)
    category = Column(String(100), nullable=True, index=True)  # Electronics, Clothing, etc.
    
    # Recipient Information
    recipient_contact_id = Column(Integer, ForeignKey("contacts.id"), nullable=True, index=True)
    recipient_name = Column(String(255), nullable=True)  # In case recipient is not a contact
    
    # Occasion and Timing
    occasion = Column(String(100), nullable=True, index=True)  # Birthday, Christmas, Anniversary, etc.
    occasion_date = Column(Date, nullable=True, index=True)
    
    # Financial Information
    budget_amount = Column(Numeric(10, 2), nullable=True)
    actual_amount = Column(Numeric(10, 2), nullable=True)
    currency = Column(String(3), nullable=False, default='USD')
    
    # Status and Priority
    status = Column(String(20), nullable=False, default=GiftStatus.IDEA.value, index=True)
    priority = Column(Integer, nullable=False, default=GiftPriority.MEDIUM.value, index=True)
    
    # Purchase Information
    store_name = Column(String(255), nullable=True)
    purchase_url = Column(Text, nullable=True)
    purchase_date = Column(Date, nullable=True)
    
    # Gift Details (JSON format for flexibility)
    # Example: {"size": "M", "color": "blue", "brand": "Nike", "model": "Air Max"}
    gift_details = Column(JSON, nullable=False, default=dict)
    
    # Tracking Information
    tracking_number = Column(String(255), nullable=True)
    delivery_date = Column(Date, nullable=True)
    
    # Notes and References
    notes = Column(Text)
    image_url = Column(Text, nullable=True)
    
    # Reminders and Notifications (JSON format)
    # Example: {"purchase_reminder": "2024-12-15", "wrap_reminder": "2024-12-20"}
    reminder_dates = Column(JSON, nullable=False, default=dict)
    
    # Integration References
    task_id = Column(Integer, nullable=True)  # Reference to shopping task
    transaction_id = Column(Integer, nullable=True)  # Reference to financial transaction
    
    # Status
    is_active = Column(Boolean, nullable=False, default=True, index=True)
    is_surprise = Column(Boolean, nullable=False, default=False)
    
    # Timestamps
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())
    
    # Relationships - using lambda to defer resolution
    tenant = relationship(lambda: Tenant, back_populates="gifts")
    user = relationship(lambda: User, back_populates="gifts")
    recipient_contact = relationship(lambda: Contact, back_populates="gifts_received")


class GiftIdea(Base):
    __tablename__ = "gift_ideas"

    id = Column(Integer, primary_key=True, index=True)
    tenant_id = Column(Integer, ForeignKey("tenants.id"), nullable=False, index=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False, index=True)
    
    # Basic Information
    title = Column(String(255), nullable=False)
    description = Column(Text)
    category = Column(String(100), nullable=True, index=True)
    
    # Target Information
    target_contact_id = Column(Integer, ForeignKey("contacts.id"), nullable=True, index=True)
    target_demographic = Column(String(100), nullable=True)  # "teenage girl", "elderly man", etc.
    
    # Occasion Context
    suitable_occasions = Column(JSON, nullable=False, default=list)  # ["birthday", "graduation"]
    
    # Financial Information
    price_range_min = Column(Numeric(10, 2), nullable=True)
    price_range_max = Column(Numeric(10, 2), nullable=True)
    currency = Column(String(3), nullable=False, default='USD')
    
    # Idea Details (JSON format for flexibility)
    # Example: {"style": "modern", "interests": ["technology", "gaming"], "preferences": {"color": "blue"}}
    idea_details = Column(JSON, nullable=False, default=dict)
    
    # Sources and References
    source_url = Column(Text, nullable=True)
    source_description = Column(Text, nullable=True)
    image_url = Column(Text, nullable=True)
    
    # Rating and Feedback
    rating = Column(Integer, nullable=True)  # 1-5 star rating
    notes = Column(Text)
    
    # Tracking
    times_gifted = Column(Integer, nullable=False, default=0)  # How many times this idea was used
    last_gifted_date = Column(Date, nullable=True)
    
    # Tags for searchability (JSON format)
    tags = Column(JSON, nullable=False, default=list)
    
    # Status
    is_active = Column(Boolean, nullable=False, default=True, index=True)
    is_favorite = Column(Boolean, nullable=False, default=False, index=True)
    
    # Timestamps
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())
    
    # Relationships - using lambda to defer resolution
    tenant = relationship(lambda: Tenant, back_populates="gift_ideas")
    user = relationship(lambda: User, back_populates="gift_ideas")
    target_contact = relationship(lambda: Contact, back_populates="gift_ideas_for_them")


# Import after class definitions to avoid circular imports
from app.models.tenant import Tenant
from app.models.user import User
from app.models.contact import Contact